util/top_gen/generator.py

```diff
@@ -3,322 +3,359 @@
+                        "input ", width, name + "_i", instances
+                    )
+                case Direction.INPUT:
+                    yield PinmuxIoToBlock(
+                        "output", width, name + "_o", instances
+                    )
+                case Direction.INOUT:
+                    yield PinmuxIoToBlock(
+                        "input ", width, name + "_i", instances
+                    )
+                    yield PinmuxIoToBlock(
+                        "input ", width, name + "_en_i", instances
+                    )
+                    yield PinmuxIoToBlock(
+                        "output", width, name + "_o", instances
+                    )
 
 
-class OutputPin(NamedTuple):
-    pin_name: str
-    idx_str: str
-    """If pin array, array index after underscore e.g. '_1'"""
-    idx_alt: str
-    """If pin array, array index surrounded by square brackets e.g. '[1]'"""
-    connected_block_io: list[BlockIoInfo]
+def generate_top(config: TopConfig) -> None:
+    """Generate a top from a top configuration."""
 
+    block_ios = list(flatten_block_ios(config.blocks))
+    pins = list(flatten_pins(config.pins, block_ios))
 
-def output_pins_iter(
-    config: TopConfig, pin_to_block_outputs: PinToBlockOutputMap
-) -> Iterator[OutputPin]:
-    for pin in config.pins:
-        if outputs := pin_to_block_outputs.get(pin.name):
-            if pin.length is None:
-                yield OutputPin(pin.name, "", "", outputs)
-            else:
-                assert pin.length == len(
-                    outputs
-                ), f"Arrayed pin '{pin.name}' must have complete mapping"
-                for i in range(pin.length):
-                    yield OutputPin(pin.name, f"_{i}", f"[{i}]", [outputs[i]])
+    block_io_to_pins = block_io_to_pin_map(block_ios, pins)
 
+    output_block_ios = list(output_block_ios_iter(block_ios, block_io_to_pins))
 
-def generate_top(config: TopConfig) -> None:
-    """Generate a top from a top configuration."""
+    def legacy_ordering(input_pin: OutputBlockIo) -> tuple[int, int, int]:
+        uid = input_pin.block_io.uid
+        io_index = 0 if uid.io_index is None else uid.io_index
 
-    pinmux_ios_to_blocks = list(ios_to_blocks_iter(config))
-    pinmux_ios_to_pins = list(ios_to_pins_iter(config))
+        block_ordering = ("uart", "spi", "gpio", "i2c")
+        block_type_idx = next(
+            idx for idx, name in enumerate(block_ordering) if uid.block == name
+        )
 
-    block_connections = block_connections_map(config)
-    pin_to_block_outputs = pin_to_block_output_map(config, block_connections)
+        return (block_type_idx, io_index, uid.instance)
 
-    input_pins = list(input_pins_iter(config, block_connections))
-    output_pins = list(output_pins_iter(config, pin_to_block_outputs))
-    inout_pins = list(
-        inout_pins_iter(config, block_connections, pin_to_block_outputs)
-    )
+    output_block_ios.sort(key=legacy_ordering)
```

Comment:
Is this so that we can change the ordering later?

Reply:
Yep